### DISCUSSION TITLE
[Typescript] How to use dynamic field names with react-hook-form components?

### DISCUSSION THREAD

**Discussion by LuisRizo** [react-hook-form/react-hook-form#6665] [Q&A]
I have been spending a good day or two looking into how to support dynamic field names in a single custom component that allows the parent component to define the field names as props. My code works with pure javascript, but I have been trying to get typescript support as well (without using `any`), and although I feel like I have made good progress, I am stuck with the problem below:

Description of the problem in a few words:

- I have an input component (uses `useFormContext`) that has multiple inputs
- The `name` for each input can be defined through a prop
- I am setting each of these names' types separately through generics like `Name extends string` (Hoping I can improve this too later on)
- Calling `register`, `setValue`, `watch` or indexing `errors` with such generic is failing
- There is no casting I can do to fix this issue

POTENTIAL FIX:

- I could cast all of my `name` props to be `Path<CustomFieldValues>` but `Path` is not listed here: https://react-hook-form.com/ts. Thus using it in my code will cause Typescript compilation to fail.

The error I'm getting in my code:

```
Argument of type 'AddressLine1Name' is not assignable to parameter of type 'Path<FullAddressSchema<AddressLine1Name, AddressLine2Name, CityName, StateName, ZipcodeName, CountryName>>'.
  Type 'string' is not assignable to type 'Path<FullAddressSchema<AddressLine1Name, AddressLine2Name, CityName, StateName, ZipcodeName, CountryName>>'.
```

![image](https://user-images.githubusercontent.com/25105030/135171408-eb269887-1893-4316-bbbc-1a9d416b83a3.png)

Note that this is a Google Places Input component, so it's multiple user-modifiable inputs that can be added to a form using react-hook-form.

I have been playing around with the types on a Typescript Playground and was able to reproduce this error.

[Here's the playground example](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAE2AYwDYEMrDjAnmbABSgEsBbEmEgN2wF44A7AV1VTgB85nGkAzEo2AJOcAM4xSjAOaiWZAEbAoohRAipg6RqLG5FG1SWmCYAbgBQSNJmx4CcAJJiAKszCaAPC7igYwXjE4ACUtBAhGVFwAQSgodFxPbVwAPhS4BnklFT8AhCCXAG0Ack0ZGAALYoBdOAB+OD50VDFsAC4cKGZgS2sMLBx8bDcPYABpYESfXMCQsIio2PjE5LSMuABREDRmJE8Aa0mIPjgXABo4Q9xjuGTC6pTelH67IbglhIncdazlJ5sBvYiOhKo4yB5PGNfCB-LMJFJZFxflALgA1dIMVHQ2H5ODEciUGjYBoAAwAJABvMYAXxJcA65KptNEjJpADpKYQQRVPOjaZYgXjud4MadsXkgqF0OFIjE4glPII+Mo4Oj6k5XO4vC50jNcZJuuqKRY4Ka4IUoYJTlrxpMRdUALR1Dpc0Hg1CQuAAMnEkkE0guRTGD0s1MKI00X3t9KFbohH1wXzR6Q6xrN5stOiuNxcjudsYqYIhUJ98P9gYtIYsYezJ1zlgsgoAYqxUNEEAgsGIxABlZAVYBkdCeE3vDtdsQAGUEwAAjAA5dBkbB6oJlmRnCwAei3ZvbneA3enQgATIvl+K4X6N9vd6aAMKUXDnlcwiW+hGbndmnswEHAF9L1xdcA1vM0AC0SDAZAICQQDVw-cswIfCAeEkZ8l1fHE12vaQLFFNMMzgK19wnY85xfPMOhA0MGwsUxlCaZBsGiWdZ1wFs2FIw8xGICAwDEEdTW4o8ZwXTCgJwz9kLHA9RNPeC3yvaTvwfJ9FOwxCb1UuBf3-DT3xAr87zgSDoNggCJIQoyZPvVDGHQgzlP9fC4EI6U5KnMSXw6ESvKEcTl0sHSPLImcz0w3zxx48iIqCmTkHUyK4EfPAX2CkyJH05K9P8dKZIALygmC4OSsySss+KdJgtCoAw5cOjs2r6p6asLAsPgeGQKgIiYTDPD88jAqwwzcOMvdovk4A4pG5ztJM1KWskrTQJ03LKtm4Cxpk8qLKcraVIW+zHKspSDpclIAAowCgfixF8tiONbPy+IEoTZLCgKXxkwbwpfca1LSzCAd0v88uBnbir2iGdKahy6u+lIAEoOk4ttJt7ftB2HX6vswsDcem-6wMW4nv3Wsnd120rl3GuGTuXVzCNNGDGAkNzR3TU1Qpi7yJIYYoeam2diluIJCeGzcubNb8hf8on+bgQWMdi0X0HFlW-oh6XTW-RKgYvAX9dwNWglJ7Wde-LLwcNpXreAU3Qey2nOa578ivMmn6CVj2Ksd6mNqly3dxq+GlqN466sd+mEfx9NqXWG67ssdNWbEDRgDZVAIGkS65aGl8kZTmWQ4idPNCznO881hTMKL129bLjPK9z43C+L3XS7Z5vs9z+324brvy8z3vLt96Hl3r9NG+7ivR9DhngHr120-gGAXWFNG-L7Ach3e6WJe+nW4G-QmZouc3aadm3gAuAP-pSyOWtd9M1gF-OxMd11C3dAaa4o4GK1HjtXTDANk-ZMDRBgCAs0ace5VxgMvdMWAYDMCgDoZmXNCgfzxsuaoHRijFCDtLb82D-4zXwUrIhL8S7mjbphShhDiFu13IUfuDCCHUOPqQ8eXtGFcODnQp+lFOHMITurOAW8MY72xn-TyBc46CLPsTbhu5L63xoZ3a+gdNEnyplDL2zDp5qOEYo6WwDqTtUYJhS6mCcEAIakrexItmGy3IT5Jx5CBFaPoY44oxtvF6N9M7YABD7aBPdgYjaBDeEO1caXZqHj-FP2odSJGbJnFmCAA)

Error in playground:
![image](https://user-images.githubusercontent.com/25105030/135170762-d96c9861-9364-42ba-b1d4-05891cfb24e5.png)

I copied the problematic types used in the latest (7.16.1) version of react-hook-form inside the playground.

I also have issues accessing the `errors.fieldName.message` but I believe these two might be related so once I have a working solution for calling `register(fieldName)`, then I believe this one should work as well, but screenshot just in case it's helpful

![image](https://user-images.githubusercontent.com/25105030/135171198-fd607447-8873-42a7-9d6d-aff2047adf02.png)

**Comment by Admiralfeb** [comment]
I've been working on something similar with `useController`. I'm having difficulty getting it to accept a generic type (The interface/type passed to `TFieldValues` works as long as it is a defined type and not a generic) for a wrapper component for FormTextField, but maybe it can help you.

Here's my form component code:
```ts
import {
  Control,
  FieldValues,
  Path,
  RegisterOptions,
  useController,
} from 'react-hook-form';

interface USCControllerProps<
  TFieldValues extends FieldValues,
  TName extends Path<TFieldValues> = Path<TFieldValues>
> {
  label: string;
  name: TName;
  // eslint-disable-next-line @typescript-eslint/ban-types
  control: Control<TFieldValues, object>;
  rules?: RegisterOptions<TFieldValues, TName>;
}

interface FormTextFieldProps<
  TFieldValues extends FieldValues,
  TName extends Path<TFieldValues> = Path<TFieldValues>
> extends USCControllerProps<TFieldValues, TName> {
  multiline?: boolean;
}

export const FormTextField = <TFieldValues extends FieldValues>({
  label,
  name,
  control,
  rules,
  multiline,
}: FormTextFieldProps<TFieldValues>) => {
  const {
    field: { ref, ...inputProps },
    fieldState: { error },
  } = useController({ name, control, rules });
  return (
    <TextFieldwM1
      label={label}
      {...inputProps}
      inputRef={ref}
      fullWidth
      multiline={multiline}
      variant="standard"
      error={error !== undefined}
      helperText={error?.message}
    />
  );
};
```

Here's the implementation:
```ts
const CmdrNameField = ({
  control,
}: {
  // eslint-disable-next-line @typescript-eslint/ban-types
  control: Control<Omit<IAmbassador, '_id'>, object>;
}) => {
  return (
    <FormTextField
      label="Cmdr Name"
      control={control}
      name="cmdrName"
      rules={{ required: 'This field is required' }}
    />
  );
};
```
As long as I pass an actual interface or type to the control there (`IAmbassador`) then it works. If I try to turn it into `T extends ICMDR` which IAmbassador does, it starts throwing red squiggles.

**Comment by LuisRizo** [comment]
Thanks for the response!

I believe one of the reasons your code also works is because you're importing `Path` from react-hook-form. And while that is exposed during development, I had compilation errors of that type not existing. I know this comment is about V6 but I'm getting it on V7 as well: https://github.com/react-hook-form/react-hook-form/issues/2049#issuecomment-653525998

The issues I get when I try importing types that aren't in https://react-hook-form.com/ts/
![image](https://user-images.githubusercontent.com/25105030/135669909-c75087b9-e29d-4f25-a689-722ee31d8c13.png)

Also, thanks for sending that code example! My generics look much nicer now thanks to you 😄 .
